automated-comments/go/two-fer/strings_builder_used_for_concatenation.md

```diff
@@ -0,0 +1,6 @@
+`strings.Builder` is a highly optimized way to build complex strings.
```

Comment:
We've discussed this a few times before, and it's another classic problem of how much information to give the student at what stage. This exercise is perfectly well-served by using `fmt.Sprintf`, and it's an extremely valuable piece of the standard library to know.

Using concatenation is basically a performance optimization, and I'm very cautious about introducing students to this idea at such an early stage (especially because about 99% of students tend to err on the side of performance over readability. If anything, they need to be encouraged to make their code slower, but more readable!)

That said, if you _do_ opt to go the concatenation route, `strings.Builder` is absolutely the best general tool for this. It's only in certain weird, unusual, extreme cases where you have _very_ few strings that `+` is a good option. I think that's _way_ out of scope for `two-fer`.

So I'm not even sure that we should be _disapproving_ the use of `strings.Builder`. I'd be more inclined to make it a suggestion. What do you think?

Reply:
Given what @bitfield says and discussions I've had with @iHiD about this in JS/TS and Ruby, I don't think this should be a disapproval comment either. 

You might even want to _leave it out alltogether_. This is the first exercise. If both `sprintf` and `Stringbuilder` are _reasonable_ solutions, accept both. There are plenty exercises later in the track with strinbuilding where this comment would probably make more sense. 

If I'm completely wrong, just ignore me :heart:

Reply:
I don't know anything about Go, but in the C# track we have a similar situation where there are several different ways in which strings can be concatenated. The C# track differentiates between string concatenation that an experienced C# programmer would never use when solving two-fer (e.g. using `string.Join`), and string concatenation options that are less idiomatic (e.g. `string.Format`). In the former case, we disapprove. In the latter case, we approve with a comment. I'm not sure which scenario applies here though, as I don't know Go.

Reply:
I don't think experienced Go developers would _ever_ use `strings.Builder` in the context of something like TwoFer. They'd use `fmt` or `+`, and I'm pretty sure their concern would be readability, not performance. At least the Go folks I've worked with.

Go has a culture of going as simple as possible, but no simpler. My gut instinct with Builder is to tell them that it's overkill for this exercise (no elaboration), and that experienced Go programmers would typically use `fmt` or simple concatenation with `+`.

Reply:
This comment is only shown if they used the `strings.Builder`.

So they either knew about it already as they are not new to Go. In this case they probably thought their solution is super fast, when it is not. They can handle this comment.

Or they searched for how to build a string in Go and found this (comparably) very complicated and highly optimized piece of software, and now think that string concatenation in Go is really complicated, because you have to use this weird `strings.Builder`.

In my opinion they should know either way, that `strings.Builder` is rarely used in production code because most cases when a strings needs building, `fmt` or `+` is simpler and even faster in case of `+`.

Reply:
I'm happy to make this an approval comment though, if that is fine with @kytrinyx

Reply:
@kytrinyx: Did not see your message before I answered this (still had the tab open). I simplified it to your suggestion but left out the `Experienced Go programmers` so students don't feel like we are labelling them inexperienced.

Reply:
> avoid students possibly feel like we are labelling them "inexperienced".

Good call!